Print a report on archives on startup & add an http endpoint for that

hidenori-shinohara opened the pull request:
# Description

Resolves https://github.com/stellar/stellar-core/issues/2328

With this change, the program will now print the name, server, and current ledger information on startup & as requested through the http-endpoint `archive-report`.

The output will be formatted as following:

```
2020-11-04T13:29:24.544 GBA2J [History INFO] Archive information: [name: sdftest1, server: stellar-core 15.0.1~unstable (572604757f7dd5760cb2da2b35956a83a3b34618), currentLedger: 121151]
```


# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [x] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

marta-lokhova commented on src/history/HistoryArchiveManager.cpp:
```diff
@@ -156,6 +158,31 @@ HistoryArchiveManager::selectRandomReadableHistoryArchive() const
+    auto& ws = mApp.getWorkScheduler();
+    for (auto const& archive : mArchives)
+    {
+        auto getHas = ws.executeWork<GetHistoryArchiveStateWork>(
```
I don't think it's a good idea to have this endpoint synchronously process all archives: downloading multiple files can take a considerable amount of time and stall the node (this will have especially negative consequences for validating nodes). In the past we've seen `curl` taking abnormally long amounts of time, so I think we should make this reporting mechanism safer. Specifically, we should not be using `executeWork` and switch to `scheduleWork` (this will execute the work asynchronously). To manage  GetHistoryArchiveStateWork works, we could add a small class that derives from Work and checks on the status of its children.

hidenori-shinohara replied:
I'm not sure if I did it correctly, but I did try to create a class `historyArchiveReportWork` that derives from `Work` and checks on the status of its children each of which is a `GetHistoryArchiveStateWork`. Let me know what you think!

marta-lokhova commented on src/history/HistoryArchiveManager.cpp:
```diff
@@ -156,6 +158,31 @@ HistoryArchiveManager::selectRandomReadableHistoryArchive() const
+HistoryArchiveManager::logReportOnArchiveConfig() const
+{
+    auto& ws = mApp.getWorkScheduler();
+    for (auto const& archive : mArchives)
```
I don't think you want to do reporting on the archives we publish to. So I think we want read only archives (there is already logic that computes those in `checkSensibleConfig`, so perhaps we just need to make it a member variable)

MonsieurNicolas replied:
It's fine to loop over all archives: people can still make typos in archives they publish to (this functionality is to help detect problems early, and not just when we need archives)

MonsieurNicolas commented:
While it's fine to have this be a separate command to perform this "on demand", I think we would always trigger this on startup (with the `run` command) as per the original issue.
Make sure to disable that "automatic" behavior in tests (as we create a lot of applications, this may slow things down too much).

MonsieurNicolas commented:
btw this is one of the "self-checks" that we were talking about it in https://github.com/stellar/stellar-core/issues/2295 - so we may want to factor the code so that we can add more checks over time.

marta-lokhova commented on src/history/HistoryArchiveManager.cpp:
```diff
@@ -156,6 +158,31 @@ HistoryArchiveManager::selectRandomReadableHistoryArchive() const
+        }
+        else
+        {
+            CLOG(INFO, "History") << fmt::format(
```
Actually, we already have logging for this in a different PR, so no need to log this error.

hidenori-shinohara replied:
removed!

hidenori-shinohara commented on src/history/HistoryArchiveManager.cpp:
```diff
@@ -156,6 +158,31 @@ HistoryArchiveManager::selectRandomReadableHistoryArchive() const
+    for (auto const& archive : mArchives)
+    {
+        auto getHas = ws.executeWork<GetHistoryArchiveStateWork>(
+            0, archive, "archive-report", 0);
```
Fixed!

hidenori-shinohara commented on src/history/HistoryArchiveManager.cpp:
```diff
@@ -58,6 +59,7 @@ HistoryArchiveManager::checkSensibleConfig() const
             }
         }
     }
+    logReportOnArchiveConfig();
```
moved it to `runWithConfig`

hidenori-shinohara commented:
I added the self-check command also. I haven' disabled the "automatic" behavior yet.

marta-lokhova reviewed (commented):
Getting closer! I think you can simplify things quite a bit.

hidenori-shinohara commented on src/history/HistoryArchiveManager.h:
```diff
@@ -4,6 +4,7 @@
 // under the Apache License, Version 2.0. See the COPYING file at the root
 // of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
 
+#include "history/HistoryArchiveReportWork.h"
```
Done!

hidenori-shinohara commented on src/history/HistoryArchiveReportWork.h:
```diff
@@ -0,0 +1,28 @@
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "historywork/GetHistoryArchiveStateWork.h"
```
Done!

hidenori-shinohara commented on src/history/HistoryArchiveReportWork.cpp:
```diff
@@ -0,0 +1,83 @@
+    {
+        for (auto const& archive : mArchives)
+        {
+            mGetHistoryArchiveStateWorks.insert(std::make_pair(
```
Fixed!

hidenori-shinohara commented on src/history/HistoryArchiveReportWork.cpp:
```diff
@@ -0,0 +1,83 @@
+        mAdded = true;
+    }
+
+    for (auto childIt = mGetHistoryArchiveStateWorks.begin();
```
Done! `doWork` now creates child works & checks their status. `onSuccess` and `onFailure` both call `logReports()` and print reports.

hidenori-shinohara commented on src/history/HistoryArchiveReportWork.h:
```diff
@@ -0,0 +1,28 @@
+    std::map<std::string, std::shared_ptr<GetHistoryArchiveStateWork>>
+        mGetHistoryArchiveStateWorks;
+    bool mAdded;
+    std::vector<std::shared_ptr<HistoryArchive>> mArchives;
```
done!

hidenori-shinohara commented on src/history/HistoryArchiveReportWork.cpp:
```diff
@@ -0,0 +1,83 @@
+}
+
+bool
+HistoryArchiveReportWork::done()
```
removed

hidenori-shinohara commented on src/main/ApplicationUtils.cpp:
```diff
@@ -171,6 +172,25 @@ httpCommand(std::string const& command, unsigned short port)
+void
+selfCheck(std::string const& command, Config cfg)
+{
+    if (command == "all" || command == "history")
```
Makes sense. I removed "all"

marta-lokhova replied:
I think we should just remove the COMMAND argument -- I don't see much value in having it, given that we only support one option for now.

hidenori-shinohara replied:
Fixed!

hidenori-shinohara commented:
I made sure to have the new http command be added as a default to the `COMMANDS` field when invoking the `run` command. I believe that I addressed all the comments.

marta-lokhova commented on docs/software/commands.md:
```diff
@@ -80,6 +80,9 @@ Command options can only by placed after command.
 * **sec-to-pub**:  Reads a secret key on standard input and outputs the
   corresponding public key.  Both keys are in Stellar's standard
   base-32 ASCII format.
+* **self-check <COMMAND>**: We currently one support only one command.
```
typo "one support"

hidenori-shinohara replied:
Fixed! Should have been more careful :(

marta-lokhova commented on src/history/HistoryArchiveReportWork.h:
```diff
@@ -0,0 +1,34 @@
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "history/HistoryArchive.h"
```
include this in HistoryArchiveReportWork.cpp, and forward declare in this file.

hidenori-shinohara replied:
Fixed!

marta-lokhova commented on src/history/HistoryArchiveReportWork.cpp:
```diff
@@ -0,0 +1,98 @@
+{
+    ZoneScoped;
+
+    if (!mAdded)
```
No need for member variable `mAdded`, you can simply check if `mGetHistoryArchiveStateWorks` is empty.

hidenori-shinohara replied:
Ah, yes! Removed.

marta-lokhova commented on src/history/HistoryArchiveReportWork.h:
```diff
@@ -0,0 +1,34 @@
+  protected:
+    BasicWork::State doWork() override;
+    void onSuccess() override;
+    void onFailureRaise() override;
```
You also need to implement `doReset` to reset the state in case there are retries (I know we don't retry, but it that changes in the future, this will not work properly). For example, see how it's done in `BatchWork::doReset()`. The rule is thumb is generally like this: if this work is going to retry, which member vars do we need to reset in order for the work to correctly run again?  So here you'd want to reset children works, for example.

hidenori-shinohara replied:
Got it. I added `doReset` that clears `mGetHistoryArchiveStateWorks`

marta-lokhova commented on src/history/HistoryArchiveReportWork.cpp:
```diff
@@ -0,0 +1,98 @@
+        // then allChildrenSuccessful() returns true despite the failed child.
+        // Thus we want to iterate through all the children in order to return
+        // the correct state.
+        if (WorkUtils::allSuccessful(mChildren))
```
parent work doesn't get scheduled that often (mostly it's just waiting for the downloads), so I don't think you need `mChildren`. Just create the list on the go, like we do it here: https://github.com/stellar/stellar-core/blob/c097a3fe36f108bcbb52e4995b81bfe617f3109a/src/historywork/PutSnapshotFilesWork.cpp#L70

hidenori-shinohara replied:
Got it! Fixed.

marta-lokhova commented on src/main/ApplicationUtils.cpp:
```diff
@@ -171,6 +173,27 @@ httpCommand(std::string const& command, unsigned short port)
+        std::shared_ptr<HistoryArchiveReportWork> w =
+            app->getHistoryArchiveManager().logReportOnArchiveConfig();
+        while (clock.crank(true) &&
+               w->getState() != BasicWork::State::WORK_SUCCESS &&
```
you can use `isDone` here

hidenori-shinohara replied:
Makes sense! This is much simpler :)

marta-lokhova commented on src/history/HistoryArchiveReportWork.cpp:
```diff
@@ -0,0 +1,98 @@
+
+    if (allChildrenDone())
+    {
+        // If some work fails quickly and gets cleaned up,
```
I think it would be clearer to say something like "Work class provides no guarantees on its stored children lifetime after each child completes, so we can't rely on allChildrenSuccessful, as works might have gotten cleaned up; instead, we store each child work and query its status directly". Something along those lines. What's important to mention here is that we can't rely on the internals of the Work class.

hidenori-shinohara replied:
Fixed!

MonsieurNicolas commented on src/history/HistoryArchiveReportWork.cpp:
```diff
@@ -0,0 +1,105 @@
+
+namespace stellar
+{
+HistoryArchiveReportWork::HistoryArchiveReportWork(
```
(we're fine with sequential execution)

MonsieurNicolas replied:
that way we avoid writing tests (which are missing in this PR anyways)

marta-lokhova replied:
WorkSequence stops if it encounters a failure so it wouldn't wait for the other children to finish

MonsieurNicolas replied:
I think this is fine to go with the simpler version: the check is supposed to pass 100% and is there to help detect bad configurations, so if the operator has to do a few iterations (if there are several broken archives), it's acceptable.

hidenori-shinohara replied:
Updated the code to use `WorkSequence`

MonsieurNicolas replied:
Look at how `WorkSequence` works (or how it's used elsewhere): I'm pretty sure you don't need to implement your own `doWork` if you derive from it - then you can just override the success/failure methods.

marta-lokhova replied:
@hidenori-shinohara you probably want something similar to `PublishWork` (then you don't need to check if each child is finished)

hidenori-shinohara replied:
Updated the code! Looking at `PublishWork` was helpful, and I was able to remove `doWork`.

hidenori-shinohara commented on docs/software/commands.md:
```diff
@@ -123,6 +125,10 @@ You can send commands to stellar-core via a web browser, curl, or using the --c
 command line option (see above). Most commands return their results in JSON
 format.
 
+* **archive-report**
```
Renamed the command to `self-check`. For now, I decided not to add any subcommands such as `history` or `all` per Marta's suggestion above.

hidenori-shinohara commented on src/main/ApplicationUtils.cpp:
```diff
@@ -171,6 +173,18 @@ httpCommand(std::string const& command, unsigned short port)
 }
 
+void
+selfCheck(Config cfg)
```
Updated the return value.

hidenori-shinohara commented on docs/software/commands.md:
```diff
@@ -80,6 +80,8 @@ Command options can only by placed after command.
 * **sec-to-pub**:  Reads a secret key on standard input and outputs the
   corresponding public key.  Both keys are in Stellar's standard
   base-32 ASCII format.
+* **self-check**: Perform history-related sanity checks, and it is planned
```
Fixed!

hidenori-shinohara commented on src/history/HistoryArchiveManager.cpp:
```diff
@@ -156,6 +157,13 @@ HistoryArchiveManager::selectRandomReadableHistoryArchive() const
     }
 }
 
+std::shared_ptr<HistoryArchiveReportWork>
```
Updated the method name to clarify that it schedules a work.

hidenori-shinohara commented on src/history/HistoryArchiveReportWork.cpp:
```diff
@@ -0,0 +1,105 @@
+// Copyright 2020 Stellar Development Foundation and contributors. Licensed
```
Updated it

hidenori-shinohara commented:
@MonsieurNicolas 
I addressed all the comments!

One thing I wasn't sure about - I thought that I might be able to avoid duplicating some code by defining a third `selfCheck` function (perhaps under the `util` directory), and having the two `selfCheck` functions in `CommandHandler.cpp` and `ApplicationUtils.cpp` call the third `selfCheck`.

I decided against it because the amount of duplication isn’t much, having 3 `selfCheck` functions may be more confusing and I wasn’t sure if in the future those two `selfCheck` functions will do the same operation. Let me know what you think.

hidenori-shinohara commented:
Hmm the code seems to run and pass all the tests on my computer, but Github says some checks were not successful. I'll take a look tomorrow.

hidenori-shinohara commented:
It passed the tests on GitHub

MonsieurNicolas commented on src/history/HistoryArchiveReportWork.cpp:
```diff
@@ -0,0 +1,57 @@
+    {
+        if (work->getState() != State::WORK_SUCCESS)
+        {
+            CLOG(INFO, "History")
```
this should be `ERROR` not `INFO` as we're reporting a problem

hidenori-shinohara replied:
makes sense! fixed.

MonsieurNicolas commented on Builds/VisualStudio/stellar-core.vcxproj:
```diff
@@ -391,6 +391,7 @@ exit /b 0
     <ClCompile Include="..\..\src\history\FileTransferInfo.cpp" />
     <ClCompile Include="..\..\src\history\HistoryArchive.cpp" />
     <ClCompile Include="..\..\src\history\HistoryArchiveManager.cpp" />
+    <ClCompile Include="..\..\src\history\HistoryArchiveReportWork.cpp" />
```
forgot to update the other file ".filters"

hidenori-shinohara replied:
updated

MonsieurNicolas reviewed (changes requested):
looks pretty much ready to be merged - just a couple more changes needed

MonsieurNicolas commented:
r+ 27052778ef0166d08f37959abed301056bb74e6f
